Refactor 2FA from FOF to Joomla Core

faynt0 opened the pull request:
#### Summary of Changes
- Refactored 2FA from FOF to Joomla Core
- Removed dependecies to FOF
- ~~Refactored 2FA templates to JLayouts~~ Reverted back for B/C, removed the FOF part that resolves the template paths and used JPluginHelper instead.
#### Testing Instructions
- **Step 1** Activate the Twofactor Auth plugins ( Google Auth, Yubikey) in the backend 
- **Step 2** Frontend: Go to the "Edit user profile" page and activate 2FA (Google Auth) and set up the authenticator
- **Step 2.1** Select Yubikey from the dropdown menu and it's template should be loaded ( Yubikey FOF dependecy was only related to the template)
- **Step 3** Logout and log in using your Google authenticator ( Frontend & Backend)
- **Step 4** After applying the patch Step 1-3 should work as before.

To test that custom templates still work create a "form.php" in "templates/protostar//html/plg_twofactorauth_totp" (Google Auth) with an echo() or similar and 
go to Frontend -> Edit User profile , select Google Autheticator and your form's content should be shown.

faynt0 committed:
updated totp & yubikey templating to use JLayout

faynt0 committed:
removed FOF dependeny from Yubikey

Commit:
refactored Totp class from fof package to libraries/encrypt package

faynt0 committed:
Merge remote-tracking branch 'origin/refactor_fof_to_core' into Refactor_2FA_to_Joomla_core

faynt0 committed:
Merge remote-tracking branch 'origin/Feature_Refactor2FATemplateToJLayouts' into Refactor_2FA_to_Joomla_core

# Conflicts:
#	plugins/twofactorauth/totp/totp.php

faynt0 committed:
removed obsolete constructor
cleaned phpcs errors

faynt0 committed:
cleaned phpcs errors

izharaazmi commented on plugins/twofactorauth/yubikey/yubikey.php:
```diff
@@ -121,26 +100,9 @@ public function onUserTwofactorShowConfiguration($otpConfig, $user_id = null)
-		// Stop output buffering and get the form contents
-		$html = @ob_get_clean();
+		$layout = new JLayoutFile('plugins.twofactorauth.yubikey.form');
+		$data = ['new_totp' => $new_totp];
```
>= PHP 5.4 only?

faynt0 committed:
changed array for < PHP 5.4 compatibility
moved files up a level because there are no namespaces yet
cleaned up xml files
simplified array extraction

faynt0 commented:
@izharaazmi thanks for pointing out, I changed the syntax.

izharaazmi commented:
IMO, the plugin default layouts should be placed inside the plugin itself. Same for the language files though. But since this plugin is part of core, I am not sure if this applies here.

I also believe uninstall of the plugin should remove all its elements. Certainly, except the library classes it depends on.

roland-d commented:
@izharaazmi The language files can stay where they are I believe because all other core plugins have their language files in the admin language folder.

As for removing the library classes, that is a more difficult question because you cannot delete them when you uninstall 1 plugin because the other needs it nor can you uninstall it when you delete the 2nd plugin because theoretically you could install another authentication plugin that requires it. So I would say we can leave it as-is.

izharaazmi commented:
@roland-d I said **except**...
> > Certainly, except the library classes it depends on.

I meant to say – when uninstalling the plugin **remove the default layouts too** alongwith any plugin only files . And **keep the library classes and any layout overrides** which is obvious anyway. You too are saying the same.

> > IMO, the plugin default layouts should be placed inside the plugin itself.

IMO, `/layouts` is not the best place for plugin's default layout files and uninstalling the plugin will not (and should not) remove it from there.

roland-d commented:
@izharaazmi Using the /layouts folder, we already have a layout there for the user profile plugin, so I believe it is the correct place.

As for the uninstallation, if I uninstall the profile plugin the layout is going to stay as well. The question should rather be should all Joomla core be rewritten to be fully uninstallable, perhaps but that is out-of-scope for this pull request.

izharaazmi commented:
Yeah, agreed on that. We'll talk over it some other day 👍

mbabker commented on libraries/joomla/encrypt/Base32.php:
```diff
@@ -0,0 +1,222 @@
+ *
+ * @since  1.0
+ */
+class Base32
```
This class and file name are incorrect.

mbabker commented on libraries/joomla/encrypt/Totp.php:
```diff
@@ -0,0 +1,183 @@
+ * @package  Encrypt
+ * @since    1.0
+ */
+class Totp
```
This class and file name are incorrect.

mbabker commented on plugins/twofactorauth/totp/totp.php:
```diff
@@ -44,13 +44,10 @@ class PlgTwofactorauthTotp extends JPlugin
 	public function __construct(&$subject, $config = array())
 	{
-		parent::__construct($subject, $config);
+		include_once JPATH_LIBRARIES . '/joomla/encrypt/Totp.php';
```
The fact you are manually including these files versus using the autoloader confirms my previous statements.

roland-d replied:
According to @wilsonge there is no autoloader in place for this.

mbabker replied:
EVERYTHING in `libraries/cms`, `libraries/joomla`, and `/libraries/legacy` is registered to the autoloader.  With the exception of the filesystem classes, as long as the class names and file structures follow convention, it's all autoloaded.  `JEncryptBase32` and `JEncryptTotp` with lowercased filenames would autoload.

izharaazmi replied:
>EVERYTHING in ... is registered to the autoloader.  With the exception of the filesystem classes,
@mbabker I'll take this as an opportunity to ask **why?** Is there any guidelines stopping us or just that there was no work done towards it?

mbabker replied:
It means either moving the files or renaming the classes.  At this point I wouldn't bother with it in 3.x.  Supposedly 4.0 will use namespaced code so class names will change there anyway.

wilsonge replied:
Legacy file names. The autoloader works mapping class names to folder paths and these class names can't match folder paths. Joomla 4 fixable - but not until then.

wilsonge replied:
Also RE: My comments. You 100% can autoload this. I think you have mixed up our namespacing conversations with the main Joomla autoloader ones. With a class name of JEncryptBase32 this would be autoloaded. You've tried to give it a namespaced style class name without putting it in a namespace. Which is definitely wrong :)

mbabker replied:
@izharaazmi There are a few other classes that don't fit the autoloader convention too but those get imported in the bootstrap files (like JText).  In theory it'd be possible to go a complete request cycle without invoking the filesystem classes so those aren't arbitrarily loaded during the bootup process meaning they still require `jimport()`.  Not the worst inconsistency out there, but it's what happens when you deal with implementing newer PHP features and trying to deprecate legacy logic.  Now if only someone would merge my PR adding PHP 5.3 features to the cache system... 😝

izharaazmi replied:
@mbabker

> In theory it'd be possible to go a complete request cycle without invoking the ~~filesystem~~ XXX classes

What I feel that **XXX** above fits for most of the library classes there. So none of those should be arbitrarily loaded during the bootup process. They must only be loaded on-demand only.

During the J1.6 or J1.7 age I had seen the class loader to scan the entire directory structure to build autoloadable list. I hadn't enough chance to see afterwards that whether it is still the same. But if it is, its bad.

1. Should there be any attempt towards making it autoload, or we must continue with `jimport` till 4.0?
2. Which PR of yours you referred to?

mbabker replied:
IIRC it was Joomla 1.7.3 or 2.5.0 that introduced "real" autoloading into `JLoader`.  Before that, everything required manual imports (`jimport()`, `include`, and `require`).  `jimport()` has a specific convention that dictates class names (in essence, the class name is whatever the last segment of the imported path is prepended with `J`, so `jimport('joomla.filesystem.file')` looks for `JFile`).  The other downfall with `jimport()` is that only for `joomla` keyed classes would it dictate a class name convention and only for the `libraries/joomla` directory; anything in another directory would just include the file.  Nothing in `libraries/cms` or `libraries/legacy` has ever been structured in a way where it couldn't be autoloaded so you should never see `jimport()` statements for those directories.  The "real" autoloader converts class name segments based on camelcased characters to directories in the filesystem (so using that convention it looks for `JFilesystemFile`).

Several classes were renamed to fit the real autoloader convention and other classes moved in the filesystem to fit the loader.  With some exceptions.  The entire filesystem package, the "base" package (`JAdapter*` classes), `JBrowser`, `JRoute`, `JText`, `JAuthentication`, and `JPlatform` are the only classes left that break the autoloader rules and require some form of manual import still.  `JLoader` technically isn't autoloadable either, but it's kinda hard to autoload the class defining your autoloader.

mbabker commented on libraries/joomla/encrypt/Base32.php:
```diff
@@ -0,0 +1,222 @@
+ *
+ * @copyright   Copyright (C) 2005 - 2016 Open Source Matters, Inc. All rights reserved.
+ * @license     GNU General Public License version 2 or later; see LICENSE
+ */
```
Is this file a straight copy/paste out of the FOF library?  If so, proper attribution must be given.

mbabker commented on libraries/joomla/encrypt/Totp.php:
```diff
@@ -0,0 +1,183 @@
+ *
+ * @copyright   Copyright (C) 2005 - 2016 Open Source Matters, Inc. All rights reserved.
+ * @license     GNU General Public License version 2 or later; see LICENSE
+ */
```
Is this file a straight copy/paste out of the FOF library?  If so, proper attribution must be given.

mbabker commented on plugins/twofactorauth/yubikey/yubikey.php:
```diff
@@ -121,26 +100,9 @@ public function onUserTwofactorShowConfiguration($otpConfig, $user_id = null)
-
-		// Stop output buffering and get the form contents
-		$html = @ob_get_clean();
+		$layout = new JLayoutFile('plugins.twofactorauth.yubikey.form');
```
The change in layout processing here is a backward compatibility break.

mbabker commented on plugins/twofactorauth/totp/totp.php:
```diff
@@ -134,24 +131,15 @@ public function onUserTwofactorShowConfiguration($otpConfig, $user_id = null)
 
 		// Start output buffering
-		@ob_start();
+		$layout = new JLayoutFile('plugins.twofactorauth.totp.form');
```
The change in layout processing here is a backward compatibility break.

mbabker commented:
> @izharaazmi Using the /layouts folder, we already have a layout there for the user profile plugin, so I believe it is the correct place.

Just because that has been done already does not make it correct or even set precedent.  The voting plugin and page navigation plugins both use `tmpl/` directories for layouts.  The stats plugin uses JLayout with a bunch of code inlined into it.  So the existing practice for the majority is to keep the code contained in the extension, NOT use the global directory.

roland-d commented:
I am sure if we put it in the tmpl/ folder the comment would have been why we don't put it in the layouts folder :/

mbabker commented:
Well right now the plugin makes use of a `tmpl/` folder so for B/C reasons whatever the logic for layout overrides is has to stay in place otherwise this cannot be accepted before 4.0 because it will cause HTML changes on sites implementing an override.

wilsonge commented:
This is a large b/c break - we have a JPlugin helper method (https://github.com/joomla/joomla-cms/blob/staging/libraries/cms/plugin/helper.php#L38) that will keep b/c with the FOF layouts and whichever way we go with that extension specific layouts should 100% go in the extension directory and NOT the global layout directory. As Michael says see how roberto did the stats plugin

Bakual commented:
Imho, as long as 3rd parties can't install stuff into the layouts folder without custom installation scripts, our core extensions shouldn't do it as well.
The layouts folder should only contain JLayouts which are reusable across various extensions. Obviously that isn't the case today but it should be.
The proper place for an extension specific layout is the tmpl folder.

Or we make a move and let 3rd parties install their JLayouts into eg /layouts/com_foo/bar.php (similar to the media folder).

izharaazmi commented:
@Bakual 👍 
However, I'd like to do it the other way around. Centralise extension specific files / layouts / media / language (already) within the extension directory as much as possible.

brianteeman commented:
Cant say I am a fan of having a file called **Totp.php** and another called **totp.php**

mbabker commented:
@brianteeman Everything in `libraries/cms`, `libraries/joomla`, and `libraries/legacy` should be all lowercased filenames since it uses Joomla's autoloader and it doesn't really account for or call for uppercased paths.

All of the namespaced code autoloaders pretty much call for title and camel cased paths.  That's how PSR-0 and PSR-4 are defined and for Joomla's namespaced loader we follow the PSR-0 convention and most of the Composer dependencies are PSR-4.

So before this can be merged the path casing will have to be corrected.

brianteeman commented:
@mbabker so my concern was correct?

On 11 August 2016 at 16:51, Michael Babker <notifications@github.com> wrote:

> @brianteeman <https://github.com/brianteeman> Everything in libraries/cms,
> libraries/joomla, and libraries/legacy should be all lowercased filenames
> since it uses Joomla's autoloader and it doesn't really account for or call
> for uppercased paths.
>
> All of the namespaced code autoloaders pretty much call for title and
> camel cased paths. That's how PSR-0 and PSR-4 are defined and for Joomla's
> namespaced loader we follow the PSR-0 convention and most of the Composer
> dependencies are PSR-4.
>
> So before this can be merged the path casing will have to be corrected.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/11553#issuecomment-239203836>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABPH8eLnUwS2bDRJNgmvNNxJifTSyF7xks5qe0UbgaJpZM4Jh5_z>
> .
>



-- 
Brian Teeman
Co-founder Joomla! and OpenSourceMatters Inc.
http://brian.teeman.net/

mbabker commented:
Yes.

brianteeman commented:
I can see files being moved to new locations and folders being moved but doesnt this have to be handled in the com_admin/script.php for sites updating

mbabker commented:
Eventually, yes.  However, as pointed out the layout stuff will need to be adjusted for B/C so that will presumably move the files back to their original location and that should cancel out any need for updating the script.php file.

faynt0 committed:
changed file and classname to match the autoload name convention

faynt0 committed:
reverted layout changes due to B/C
refactored from FOF override path to use core functions

faynt0 committed:
Added proper attribution

faynt0 commented:
First of all thanks everyone for your input, in the recent changes I got the autoloading to work like some of you mentioned. 

Due to your B/C concerns I reverted the changes to the templating but removed and replaced the FOF part that handles the custom template path

brianteeman commented:
I have tested this item :white_check_mark: successfully on 4bed2e5a5ef16c4d44978fd7d023b488ac2e1977<br /><br />applied PR and confirmed that both yubikey and google authenticator worked for site and admin login<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11553">issues.joomla.org/joomla-cms/11553</a>.</sub>

faynt0 commented:
Added small test instruction for custom templates.

mbabker commented on plugins/twofactorauth/totp/totp.php:
```diff
@@ -135,9 +114,7 @@ public function onUserTwofactorShowConfiguration($otpConfig, $user_id = null)
-
-		// Include the form.php from a template override. If none is found use the default.
-		$path = FOFPlatform::getInstance()->getTemplateOverridePath('plg_twofactorauth_totp', true);
+		$path = JPATH_THEMES . "/" . JFactory::getApplication()->getTemplate() . "/html/plg_twofactorauth_totp";
```
Use [JPluginHelper::getLayoutPath()](https://api.joomla.org/cms-3/classes/JPluginHelper.html#method_getLayoutPath).  It'll also eliminate the need for the if/else include, you just include the return from that method.  Same for the other plugin.

mbabker commented on plugins/twofactorauth/totp/tmpl/form.php:
```diff
@@ -8,6 +8,7 @@
  */
 
 defined('_JEXEC') or die;
+
```
You can remove this single line break to keep this file from being part of the PR.

mbabker commented on plugins/twofactorauth/totp/totp.xml:
```diff
@@ -12,7 +12,6 @@
 	<files>
 		<filename plugin="totp">totp.php</filename>
 		<folder>postinstall</folder>
-		<folder>tmpl</folder>
```
This line should be restored.

mbabker commented on plugins/twofactorauth/yubikey/tmpl/form.php:
```diff
@@ -8,6 +8,7 @@
  */
 
 defined('_JEXEC') or die;
+
```
Same here on removing the line break.

mbabker commented on plugins/twofactorauth/yubikey/yubikey.xml:
```diff
@@ -11,7 +11,6 @@
 	<description>PLG_TWOFACTORAUTH_YUBIKEY_XML_DESCRIPTION</description>
 	<files>
 		<filename plugin="yubikey">yubikey.php</filename>
-		<folder>tmpl</folder>
```
Restore this line.

faynt0 committed:
Use JPluginHelper to resolve the template path
Fixed xml files

izharaazmi commented on plugins/twofactorauth/yubikey/yubikey.php:
```diff
@@ -125,19 +104,7 @@ public function onUserTwofactorShowConfiguration($otpConfig, $user_id = null)
-		{
-			include_once __DIR__ . '/tmpl/form.php';
-		}
+		include_once JPluginHelper::getLayoutPath('twofactorauth', 'yubikey', "form");
```
I think layouts should not be `include_once`'d. Please verify this and amend if needed.

roland-d replied:
@izharaazmi Why would we need include here? The form is not going to change during runtime is it?

izharaazmi replied:
1. In this specific case, it may not apply but it is a general practice to 'require' a layout file. An 'include' may also be preferred sometimes.
2. If we call this plugin twice in a request cycle, 2nd will output nothing. Again for this case this may not be an issue, but *_once are costly anyway.

roland-d replied:
It has always been a require_once as you can see in the old code and hasn't given a problem so far.

izharaazmi replied:
> ... hasn't given a problem so far.

is not a good reason to keep doing something. To me *_once should be used only when any attempt to include the file for a second time may cause an issue such as for class definitions or function declarations etc.

Any other code should be able to load as many time as needed. See the various Joomla modules for example:

`mod_footer` would normally never be called twice but `modules/mod_footer/mod_footer.php:37` still calls it using `require`. For me it is the correct approach. No need to pose an unnecessary job to check against `included_files` to the interpreter.

andrepereiradasilva commented:
if the intention is to remove fof 2 in joomla 4.0 this should be rebased to 4.0 branch

Bakual commented:
Why? this could already be done in J3 fine.

mbabker commented:
Is this still being worked on?

franz-wohlkoenig commented:
If this PR get no Response, it will be closed at 22th June 2017.

roland-d commented:
I guess the students have given up on this. @icampus any idea?

faynt0 commented:
I kind of fail to see what is left to be done here, is it about the `include` vs. `include_once` debate?

roland-d commented:
@faynt0 At least the PR needs to be updated so it is in-sync with the staging branch.

faynt0 committed:
Merge branch 'staging' into Refactor_2FA_to_Joomla_core

franz-wohlkoenig commented:
@roland-d can you please have a Look at Status of this PR? <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11553">issues.joomla.org/tracker/joomla-cms/11553</a>.</sub>

roland-d commented:
@franz-wohlkoenig As far as I can see this needs testing as @faynt0 already updated it with the staging branch.

franz-wohlkoenig commented:
thanks @roland-d

franz-wohlkoenig commented:
@brianteeman can you please test again?

roland-d commented:
@faynt0 could you please bring this up-to-date with the J4 codebase? Pretty please :) Ping @wilsonge as well if you have questions.

wilsonge commented:
We need to update the Encrypt code with the latest version. change the branch to be against J4 and I promise I will get this merged!

wilsonge commented:
OK i've ported this in https://github.com/joomla/joomla-cms/pull/17687
